micro-rdk/src/common/ota.rs

```diff
@@ -0,0 +1,359 @@
+    InvalidImageSize(usize),
+    #[error("{0}")]
+    Other(String),
+}
+
+#[cfg(feature = "esp32")]
+type OtaConnector = crate::esp32::tcp::Esp32H2Connector;
+#[cfg(not(feature = "esp32"))]
+type OtaConnector = crate::native::tcp::NativeH2Connector;
+
+pub(crate) struct OtaService {
+    exec: Executor,
+    connector: OtaConnector,
+    url: String,
+}
+
+impl OtaService {
+    pub(crate) fn from_config(
+        ota_config: &ServiceConfig,
+        cert: TlsCertificate,
+        exec: Executor,
+    ) -> Result<Self, OtaError> {
+        // TODO(RSDK-9205)
+        let kind: Kind = ota_config
+            .attributes
+            .as_ref()
+            .ok_or_else(|| OtaError::ConfigError("config missing `attributes`".to_string()))?
+            .fields
+            .get("url")
+            .ok_or(OtaError::ConfigError(
+                "config missing `url` field".to_string(),
+            ))?
+            .kind
+            .as_ref()
+            .ok_or(OtaError::ConfigError(
+                "failed to get inner `Value`".to_string(),
+            ))?
+            .try_into()
+            .map_err(|e: AttributeError| OtaError::ConfigError(e.to_string()))?;
+
+        let url = match kind {
+            Kind::StringValue(s) => s,
+            _ => {
+                return Err(OtaError::ConfigError(format!(
+                    "invalid url value: {:?}",
+                    kind
+                )))
+            }
+        };
+
+        let mut connector = OtaConnector::default();
+
+        connector.set_server_certificates(cert.certificate.clone(), cert.private_key.clone());
+
+        Ok(Self {
+            url,
+            connector,
+            exec,
+        })
+    }
+
+    pub(crate) async fn update(&mut self) -> Result<(), OtaError> {
+        let mut uri = self
+            .url
+            .parse::<hyper::Uri>()
+            .map_err(|_| OtaError::ConfigError(format!("invalid url: {}", self.url)))?;
+        if uri.port().is_none() {
+            if uri.scheme_str() != Some("https") {
+                log::error!("no port found and not https");
+            }
+
+            let mut auth = uri.authority().expect("str").to_string();
+            auth.push_str(":443");
+            let mut parts = uri.into_parts();
+            parts.authority = Some(
+                auth.parse()
+                    .map_err(|_| OtaError::Other("failed to parse authority".to_string()))?,
+            );
+            uri = hyper::Uri::from_parts(parts).map_err(|e| OtaError::Other(e.to_string()))?;
+        };
+        let io = self
+            .connector
+            .connect_to(&uri)
+            .unwrap()
+            .await
+            .map_err(|e| OtaError::Other(e.to_string()))?;
+
+        let (mut sender, conn) = {
+            http2::Builder::new(self.exec.clone())
+                .keep_alive_interval(Some(std::time::Duration::from_secs(120)))
+                .keep_alive_timeout(std::time::Duration::from_secs(10))
```

Comment:
Is the minimum amount of time we can use for this timeout? If a smaller timeout works, I think it would be preferred

Reply:
is there a lower limit you have in mind?

Reply:
no, I figured we could arrive at this by lowering the duration until it no longer works on a relatively unstable network (probably do something more like binary search)

Reply:
Is there a reason that we know we need to adjust these from the defaults?

Reply:
confirmed working without setting `keep_alive_interval` and using default timeout (20 seconds).